src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    // NOTE: this type depends on Dev15 assemblies, which is why the type is in EditorFeatures.Next. But, that library
+    // is rehostable and once we move .editorconfig support fully through the system, it should be moved to Workspaces
+    // or perhaps even lower.
+    internal sealed class EditorConfigDocumentOptionsProvider : IDocumentOptionsProvider
+    {
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// The map of cached contexts for currently open documents.
+        /// </summary>
+        private readonly Dictionary<DocumentId, Task<ICodingConventionContext>> _openDocumentContexts = new Dictionary<DocumentId, Task<ICodingConventionContext>>();
+
+        private readonly ICodingConventionsManager _codingConventionsManager;
+
+        internal EditorConfigDocumentOptionsProvider(Workspace workspace)
+        {
+            _codingConventionsManager = CodingConventionsManagerFactory.CreateCodingConventionsManager();
+
+            workspace.DocumentOpened += Workspace_DocumentOpened;
+            workspace.DocumentClosed += Workspace_DocumentClosed;
+        }
+
+        private void Workspace_DocumentClosed(object sender, DocumentEventArgs e)
+        {
+            lock (_gate)
+            {
+                Task<ICodingConventionContext> contextTask;
+
+                if (_openDocumentContexts.TryGetValue(e.Document.Id, out contextTask))
+                {
+                    _openDocumentContexts.Remove(e.Document.Id);
+
+                    // Ensure we dispose the context, which we'll do asynchronously
+                    contextTask.ContinueWith(t => t.Result.Dispose(), CancellationToken.None, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
```

Comment:
It seemed like some unnecessary complexity, also strangeness could then begin if the task we're awaiting later then gets that cancelled out underneath it. Yeah, we could use AsyncLazy and whatnot but I don't anticipate a problem here.